services/graphite/service.go

```diff
@@ -23,16 +23,49 @@ const (
 	leaderWaitTimeout = 30 * time.Second
 )
 
-// expvar stats maintained by the Graphite package.
+// Initialize the graphite stats and diags
+func init() {
+	tcpConnections = make(map[string]*tcpConnectionDiag)
+}
+
+// Package-level tracking of connections for diagnostics.
 var monitorOnce sync.Once
-var statMap = expvar.NewMap("graphite")
-var statMapTCP = expvar.NewMap("tcp")
-var statMapUDP = expvar.NewMap("udp")
 
-// Build the graphite expvar hierarchy.
-func init() {
-	statMap.Set("tcp", statMapTCP)
-	statMap.Set("udp", statMapUDP)
+type tcpConnectionDiag struct {
+	conn        net.Conn
+	connectTime time.Time
+}
+
+var tcpConnectionsMu sync.Mutex
+var tcpConnections map[string]*tcpConnectionDiag
```

Comment:
Mostly because I went for a simple diagnostics display, something like you
get when you run netstat - a single table showing all connections to
Graphite, not broken down by individual Graphite inputs. It's clear anyway
because the table shows source and dest IP.

I did not break stats down by connection because someone that uses a lot of
different connections that come and might start creating a lot of stats
noise - and series.

On Friday, September 4, 2015, dgnorton <notifications@github.com> wrote:

> In services/graphite/service.go
> <https://github.com/influxdb/influxdb/pull/3969#discussion_r38747327>:
>
> >
> > -// Build the graphite expvar hierarchy.
> > -func init() {
> > -	statMap.Set("tcp", statMapTCP)
> > -	statMap.Set("udp", statMapUDP)
> > +type tcpConnectionDiag struct {
> > +	conn        net.Conn
> > +	connectTime time.Time
> > +}
> > +
> > +var tcpConnectionsMu sync.Mutex
> > +var tcpConnections map[string]*tcpConnectionDiag
>
> Why not track connections in the service also?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/influxdb/influxdb/pull/3969/files#r38747327>.
>